patch: Add _update_bundle.yaml

lucasgameiroborges opened the pull request:
This PR aims to condense both previous patches #237 and #235 in a single, experimental `_update_bundle.yaml` workflow that will pin both snap and rock resource versions, in accordance to what was previously discussed:

- For rocks: Roughly follows John Meinel's proposed format, with `oci-*` lines commented in order to preserve compatibility with `juju deploy`.  See the following PR commit: https://github.com/canonical/postgresql-k8s-bundle/commit/2683a2d91004546787e2be8fc48bc81dc01d3a77
- For snaps: follows existing [store proxy standard](https://documentation.ubuntu.com/snap-store-proxy/en/airgap-charmhub/#export-snap-resources). See the following PR commit: https://github.com/canonical/postgresql-bundle/commit/b223dc3c8d176403cea67c441f5dadc99813b4f8

See also test runs on:
 - mysql bundle VM: https://github.com/canonical/mysql-bundle/pull/77/commits/612d5316706cadc45073290c1a55034617ccadca
  - mysql bundle K8s: https://github.com/canonical/mysql-k8s-bundle/pull/81/commits/018fae7197d47e049989390afb04e3f4ed510f48

Important info:
 - `storage-admin` currently has 2 blocking issues: 
    - Pinning of charm revisions is being ignored: https://bugs.launchpad.net/snapstore-client/+bug/2083876
    - Charm resource (e.g. rocks) version pinning not supported: https://bugs.launchpad.net/snapstore-client/+bug/2083878 
 - This is `amd64` only. An eventual patch for `arm64` should be straightforward.
 - This is specific to SQL bundles and their associated charms. This is the case because snap pinning requires specific logic per-charm, see point below.
 - This fetches snap revisions by individual parsing of each charm's source code (not by deploying the bundle, like on previous patch)

lucasgameiroborges commented on .github/workflows/_update_bundle.md:
thanks!

lucasgameiroborges commented on python/cli/data_platform_workflows_cli/update_bundle.py:
```diff
@@ -24,53 +37,245 @@ def get_ubuntu_version(series: str) -> str:
-    response.raise_for_status()
-    channel_map = response.json()["channel-map"]
-    track, risk = charm_channel.split("/")
+    return response.json()["channel-map"], response.json()["default-release"].get("resources", [])
```
tests worked fine, but apparently this may lead to issues yes - TIL! reverting

lucasgameiroborges commented on .github/workflows/_update_bundle.yaml:
```diff
@@ -0,0 +1,87 @@
+      path-to-snaps-file:
+        description: Relative path to snaps yaml file from repository directory
+        required: false
+        type: string
```
Done, LMK if it's what you had in mind. BTW, what question about k8s?

lucasgameiroborges replied:
Oh okay, I'll close this then, so it doesn't distract from the remaining comment.

lucasgameiroborges commented on python/cli/data_platform_workflows_cli/update_bundle.py:
```diff
@@ -24,53 +37,245 @@ def get_ubuntu_version(series: str) -> str:
     return max(revisions)
 
 
+def fetch_oci_image_metadata(download_url) -> Tuple[str, str, str]:
```
agree, refactored this.

lucasgameiroborges commented on python/cli/data_platform_workflows_cli/update_bundle.py:
```diff
@@ -24,53 +37,245 @@ def get_ubuntu_version(series: str) -> str:
+
+def fetch_grafana_snaps():
+    """Fetch grafana-agent snaps information."""
+    response = requests.get("https://raw.githubusercontent.com/canonical/grafana-agent-operator/refs/heads/main/src/snap_management.py")
```
you are right, we should use `refs/tags/revXX` instead of `refs/heads/main`, thanks for pointing this out!

lucasgameiroborges commented:
I think I addressed all of @carlcsaposs-canonical 's comments - Let me know how it looks. About your previous question:

> question: do @delgod's requirements mean that we need to have a bundle for k8s that's deployable (which would require the resources to be omitted)? 

If I remember correctly, @taurus-forever suggested that we include all OCI info in commented lines so that k8s bundle files stay deployable, like this:
```
  postgresql-k8s:
    channel: 14/edge
    charm: postgresql-k8s
    constraints: arch=amd64
    revision: 404
    resources:
      postgresql-image: 164  # <== juju deploy requires that value to be either int (the rev) or str (the oci image path)
       # oci-image: ....
       # oci-password: ...
       # oci-username: ...
```

My local tests show that this format works in `juju deploy`. If everyone approves, I can adapt the PR.

lucasgameiroborges commented:
> is https://github.com/canonical/postgresql-bundle/pull/208 up-to-date? I'm not seeing the snap yaml in that example

got overwritten by the old workflow, still running, will re-run in a sec, but you can refer to the previous commit: https://github.com/canonical/postgresql-bundle/commit/53eac6a07ead599d756281d339750c00f1ba0731

lucasgameiroborges commented on python/cli/data_platform_workflows_cli/update_bundle.py:
```diff
@@ -24,53 +36,245 @@ def get_ubuntu_version(series: str) -> str:
+            yaml.dump(bundle_data, file)
+
+    if len(snaps_data["packages"]) > 0:
+        snaps_data["packages"] = remove_snap_duplicates(snaps_data["packages"])
```
great suggestion, added!

taurus-forever commented:
@lucasgameiroborges why two snap revisions [here](https://github.com/canonical/postgresql-bundle/commit/8716a3893f04f5328a7afa0c1575cea6d926eecf#diff-331fc9c29e596d215da7f364bcc6ad690f02966ef62849153a5126977e882874R5-R10) ?

re: K8s, should we comment YAML lines to make juju deploy bundle happy?

taurus-forever commented:
> I think I addressed all of @carlcsaposs-canonical 's comments - Let me know how it looks. About your previous question:
> 
> > question: do @delgod's requirements mean that we need to have a bundle for k8s that's deployable (which would require the resources to be omitted)?
> 
> If I remember correctly, @taurus-forever suggested that we include all OCI info in commented lines so that k8s bundle files stay deployable, like this:
> 
> ```
>   postgresql-k8s:
>     channel: 14/edge
>     charm: postgresql-k8s
>     constraints: arch=amd64
>     revision: 404
>     resources:
>       postgresql-image: 164  # <== juju deploy requires that value to be either int (the rev) or str (the oci image path)
>        # oci-image: ....
>        # oci-password: ...
>        # oci-username: ...
> ```
> 
> My local tests show that this format works in `juju deploy`. If everyone approves, I can adapt the PR.

Go this way for now (to make Juju happy about the bundle file).

Ping @jameinel :  should we bugreport juju about lack of revision support as YAML node? Tnx!

lucasgameiroborges commented:
K8s bundle is deployable again, all syntax-breaking`oci-*` lines are now added as comments. Also added links to the 2 blocking issues currently on `store-admin`'s side, please refer to updated PR description for all relevant links

lucasgameiroborges commented on python/cli/data_platform_workflows_cli/update_bundle.py:
```diff
@@ -239,34 +238,41 @@ def main():
-                        "oci-username": resource_data["Username"],
-                        "oci-password": resource_data["Password"],
-                    }
+                app.setdefault("resources", {})
+                app["resources"][resource["name"]] = int(resource["revision"])
```
Oh! Missed this comment before the merge, sorry :(
